### PR TITLE
BFD-1529: Populate Part B Termination Code Correctly

### PR DESCRIPTION
**Pull request by JeremyIR**
**JIRA Ticket:**
[BFD-1529](https://jira.cms.gov/browse/BFD-1529)

**User Story or Bug Summary:**

As a BFD engineer, I want the correct value for the Part B termination code in v2 coverage coded extension, so I can use that field.

---

### What Does This PR Do?

Use the correct part B termination code instead of partA.

### What Should Reviewers Watch For?

Are we using the correct partB termination code? 

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by mackec1 — APPROVED** [review]
Ship it!

**Review by mackec1 — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java [hunk 1]
```diff
@@ -184,7 +184,7 @@ private static Coverage transformPartB(MetricRegistry metricRegistry, Beneficiar
         coverage, CcwCodebookVariable.MS_CD, beneficiary.getMedicareEnrollmentStatusCode());
 
     addCoverageCodeExtension(
-        coverage, CcwCodebookVariable.B_TRM_CD, beneficiary.getPartATerminationCode());
+        coverage, CcwCodebookVariable.B_TRM_CD, beneficiary.getPartBTerminationCode());
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java:187]: We are missing a unit test and a changelog entry.

Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2.java:187]: Can we change the partATerminationCode in the test data so that it is different than partBTerminationCode? That should make the test fail with the existing code and your change here will cause that test to pass. It should also necessitate changing the tests that validate partATerminationCode.

For the api changelog I'm not sure that I agree. The only reason why the JSON didn't change is because our test data has the same values in these fields. That is not necessarily true for real data (I assume, otherwise they would be the same field) so consumers of the API will potentially see different, correct data showing up after this change. fyi @bfausett77

**Review by lsmitchell — APPROVED** [review]

**Review by mackec1 — APPROVED** [review]

**Review by mackec1 — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java [hunk 2]
```diff
@@ -699,11 +699,11 @@ static void assertPartAMatches(Beneficiary inBeneficiary, Coverage inCoverage) {
     currSegment = MedicareSegment.PART_A;
     assertNotNull(coverage);
     assertNotNull(beneficiary);
+    assertEquals("cancelled", coverage.getStatus().toCode());
 
     verifyCoverageClass("Part A");
     verifyMeta();
     verifyExtensionsPartA();
-    verifyCoverageStatus();
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java:706]: why was this removed and the direct assertion added? do we still use this method in other places where we expect the status to be active? if so adding a method parameter to the method would be more consistent

File: apps/bfd-server/dev/api-changelog.md [hunk 3]
```diff
@@ -1,5 +1,72 @@
 # API Changelog
 
+## BFD-1529: Populate Part B Termination Code Correctly
+
+For Coverage Claims, the Part A coverage termination code was being provided by the API for Part B events. 
+In the test dataset, Part A and Part B coverage termination codes are the same, 
+however in production, this will not always be the case.
+So in addition to providing the Part B termination code correctly, 
+Part A will have a different termination code than Part B
```
[the remark this answers is not in the document]
Comment (bfausett77) [reply apps/bfd-server/dev/api-changelog.md:9]: I think the information in the changelog can be even more simple.  I don't think we even need to provide an example like we are below.  The change is literally just ensuring that the part-b termination code that we have for the patient is being displayed instead of the part-a termination code.  So I would just have the API ChangeLog say:

"Previously the Part B coverage termination code was always the same as the Part A coverage termination code.  The data will now reflect the proper Part B termination code."  Thoughts?

[the remark this answers is not in the document]
Comment (bfausett77) [reply apps/bfd-server/dev/api-changelog.md:9]: @lsmitchell @christrotteradhoc

**Review by bfausett77 — DISMISSED** [review]
Looks good.  Thanks for the changes in the 'change log'.  Nice work!

File: apps/bfd-server/dev/api-changelog.md [hunk 4]
```diff
@@ -1,5 +1,10 @@
 # API Changelog
 
+## BFD-1529: Populate Part B Termination Code Correctly
+
+Previously the Part B coverage termination code was always the same as the Part A coverage termination code.
+The data will now reflect the proper Part B termination code
```
Comment (christrotteradhoc) [inline apps/bfd-server/dev/api-changelog.md:6]: nitpick -- missing a period

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java [hunk 5]
```diff
@@ -146,7 +146,7 @@ private static void verifyExtensionsPartA() {
   @Test
   public void verifyCoverageStatusPartA() {
     transformCoverage(MedicareSegment.PART_A, false);
-    verifyCoverageStatus();
+    assertEquals("cancelled", coverage.getStatus().toCode());
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java:149]: can't this be `verifyCoverageStatus("cancelled")` now ?

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java [hunk 6]
```diff
@@ -576,8 +576,18 @@ private static void verifyCodedExtension(String url, String code, String display
...
+    if (status != null && !"".equals(status)) {
+      if ("active".equals(status)) {
+        assertEquals("active", coverage.getStatus().toCode());
+      } else if ("cancelled".equals(status)) {
+        assertEquals("cancelled", coverage.getStatus().toCode());
+      } else if ("draft".equals(status)) {
+        assertEquals("draft", coverage.getStatus().toCode());
+      } else if ("entered-in-error".equals(status)) {
+        assertEquals("entered-in-error", coverage.getStatus().toCode());
+      }
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java:589]: can we just make this `assertEquals(status, coverage.getStatus().toCode())` ?

**Review by mackec1 — DISMISSED** [review]
I'll assume you'll add the 'period' to your change log entry.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java (hunk 6 above)
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java]: I was not very clear, sorry. I was thinking the the entire method body can be replaced with just one assertEquals.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java [hunk 7]
```diff
@@ -576,8 +576,18 @@ private static void verifyCodedExtension(String url, String code, String display
     assertTrue(compare.equalsDeep(ex));
   }
 
-  private static void verifyCoverageStatus() {
-    assertEquals("active", coverage.getStatus().toCode());
+  private static void verifyCoverageStatus(String status) {
+    if (status != null && !"".equals(status)) {
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java:580]: Generally I think this would be written ` if (status != null && status.length > 0)` just to be slightly easier to read. (Apache common utils also has a nice util StringUtils.isNotBlank too for doing this type of null+empty check)

**Review by lsmitchell — APPROVED** [review]
Approved pending period adjustment

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java [hunk 8]
```diff
@@ -576,8 +576,10 @@ private static void verifyCodedExtension(String url, String code, String display
     assertTrue(compare.equalsDeep(ex));
   }
 
-  private static void verifyCoverageStatus() {
-    assertEquals("active", coverage.getStatus().toCode());
+  private static void verifyCoverageStatus(String status) {
+    if (status != null && status.length() > 0) {
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java:580]: do we actually need this condition? I would think that if a test passed in null or an empty string here then this method should assert that the coverage code is such. If that is wrong then the test should not assert it as so.

The risk I see here is that if a test does pass null, this will not assert anything and silently just work.

Comment (lsmitchell) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/CoverageTransformerV2Test.java]: yup with the refactor of the method this was no longer needed

**Review by lsmitchell — APPROVED** [review]

**Review by christrotteradhoc — APPROVED** [review]
